edge-modules/api-proxy-module/src/main.rs

```diff
@@ -101,60 +112,80 @@ pub fn nginx_controller_start(
-                .context("Cannot start proxy!")?;
-
-            // Restart nginx on new config, new cert or crash.
-            let cert_reload = notify_cert_reload_api_proxy.notified();
-            let config_reload = notify_config_reload_api_proxy.notified();
-            futures::pin_mut!(cert_reload, config_reload);
-            let signal_restart_nginx = future::select(cert_reload, config_reload);
-            futures::pin_mut!(child_nginx, signal_restart_nginx);
-            let restart_proxy = future::select(child_nginx, signal_restart_nginx);
-
-            //Shutdown on ctrl+c or on signal
-            let wait_shutdown_ctrl_c = shutdown::shutdown();
-            futures::pin_mut!(wait_shutdown_ctrl_c);
-            let wait_shutdown_signal = shutdown_signal.notified();
-            futures::pin_mut!(wait_shutdown_signal);
-            let wait_shutdown = future::select(wait_shutdown_ctrl_c, wait_shutdown_signal);
-
+            let nginx_start =
+                nginx_command(proxy_name, program_path, &start_proxy_args, "start")?.fuse();
+            futures::pin_mut!(nginx_start);
             info!("Starting/Restarting API-Proxy");
-            match future::select(wait_shutdown, restart_proxy).await {
-                Either::Left(_) => {
-                    warn!("Shutting down ngxing controller!");
-                    return Ok(());
-                }
-                Either::Right((result, _)) => {
-                    match result {
-                        Either::Left(_) => {
-                            info!("Nginx crashed, restarting");
-                        }
-                        Either::Right(_) => {
-                            info!("Request to restart Nginx received");
-                        }
-                    };
+
+            loop {
+                //Shutdown nginx on ctrl_c or signal
+                let wait_shutdown_ctrl_c = shutdown::shutdown().fuse();
+                let wait_shutdown_signal = shutdown_signal.notified().fuse();
+
+                // Restart nginx on new config, new cert or crash.
+                let cert_reload = notify_cert_reload_api_proxy.notified().fuse();
+                let config_reload = notify_config_reload_api_proxy.notified().fuse();
+
+                futures::pin_mut!(
+                    wait_shutdown_ctrl_c,
+                    wait_shutdown_signal,
+                    cert_reload,
+                    config_reload
+                );
+
+                // Bug in clippy, not using mut mut here
+                #[allow(clippy::mut_mut)]
+                let proxy_command = select! {
+                    _ = wait_shutdown_ctrl_c => NginxCommands::Stop,
+                    _ = wait_shutdown_signal => NginxCommands::Stop,
+                    _ = cert_reload => NginxCommands::Reload,
+                    _ = config_reload => NginxCommands::Reload,
+                    _ = nginx_start => NginxCommands::Restart,
```

Comment:
When does this return? When nginx crashes?

Reply:
No when nginx crashes it just restarts it.
It returns on  NginxCommands::Stop. During a signal or ctrl+c

Reply:
I am not sure I follow. My questions is - when does nginx_start future return, resulting in NginxCommands;:Restart being executed

Reply:
Ah I see. Yes, what you said first is correct. The future returns what nginx crashes. I though you meant return from the function.